File: java/datastax-v4/spring/src/main/java/com/aws/ssa/keyspaces/retry/AmazonKeyspacesRetryPolicy.java
package com.aws.ssa.keyspaces.retry;

import com.datastax.oss.driver.api.core.ConsistencyLevel;
import com.datastax.oss.driver.api.core.config.DriverExecutionProfile;
import com.datastax.oss.driver.api.core.context.DriverContext;
import com.datastax.oss.driver.api.core.retry.RetryDecision;
import com.datastax.oss.driver.api.core.retry.RetryPolicy;
import com.datastax.oss.driver.api.core.servererrors.*;
import com.datastax.oss.driver.api.core.session.Request;
import com.datastax.oss.driver.shaded.guava.common.annotations.VisibleForTesting;
import edu.umd.cs.findbugs.annotations.NonNull;
import net.jcip.annotations.ThreadSafe;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;


/**
 * This is a conservative retry policy adapted for the Amazon Keyspaces Service.
 * It allows for a configurable number of attempts, but by default the number of attempts is {@value AmazonKeyspacesRetryOption#DEFAULT_KEYSPACES_RETRY_MAX_ATTEMPTS}
 * <p>
 * This policy will either reattempt request on the same host or rethrow the exception to the calling thread. The main difference between
 * this policy from the original {@link com.datastax.oss.driver.internal.core.retry.DefaultRetryPolicy} is that the {@link AmazonKeyspacesRetryPolicy} will call {@link RetryDecision#RETRY_SAME} instead of {@link RetryDecision#RETRY_NEXT}
 * <p>
 * In Amazon Keyspaces, it's likely that {@link WriteTimeoutException} or {@link ReadTimeoutException} is the result of exceeding current table
 * capacity. Learn more about Amazon Keyspaces capacity here: @see <a href="https://docs.aws.amazon.com/keyspaces/latest/devguide/ReadWriteCapacityMode.html">Amazon Keyspaces CapacityModes</a>.
 * In most cases you should allow for small number of retries, and handle the exception in your application threads.
 *
 * <p>To activate this policy, modify the {@code advanced.retry-policy} section in the driver
 * configuration, for example:
 *
 * <pre>
 * datastax-java-driver {
 *   advanced.retry-policy {
 *     class = com.aws.ssa.keyspaces.retry.AmazonKeyspacesRetryPolicy
 *     max-attempts = 2
 *   }
 * }
 * </pre>
 */

@ThreadSafe
public class AmazonKeyspacesRetryPolicy implements RetryPolicy {


    private static final Logger LOG = LoggerFactory.getLogger(AmazonKeyspacesRetryPolicy.class);
    @VisibleForTesting
    public static final String RETRYING_ON_READ_TIMEOUT = "[{}] Retrying on read timeout on same host (consistency: {}, required responses: {}, received responses: {}, data retrieved: {}, retries: {})";
    @VisibleForTesting
    public static final String RETRYING_ON_WRITE_TIMEOUT = "[{}] Retrying on write timeout on same host (consistency: {}, write type: {}, required acknowledgments: {}, received acknowledgments: {}, retries: {})";
    @VisibleForTesting
    public static final String RETRYING_ON_UNAVAILABLE = "[{}] Retrying on unavailable exception on next host (consistency: {}, required replica: {}, alive replica: {}, retries: {})";
    @VisibleForTesting
    public static final String RETRYING_ON_ABORTED = "[{}] Retrying on aborted request on next host (retries: {})";
    @VisibleForTesting
    public static final String RETRYING_ON_ERROR = "[{}] Retrying on node error on next host (retries: {})";

    private final String logPrefix;

    private final Integer maxRetryCount;


    public AmazonKeyspacesRetryPolicy(DriverContext context) {
        this(context, context.getConfig().getDefaultProfile().getName());
    }

    public AmazonKeyspacesRetryPolicy(DriverContext context, Integer maxRetryCount) {

        String profileName = context.getConfig().getDefaultProfile().getName();

        this.maxRetryCount = maxRetryCount;

        this.logPrefix = (context != null ? context.getSessionName() : null) + "|" + profileName;
    }

    public AmazonKeyspacesRetryPolicy(DriverContext context, String profileName) {
        DriverExecutionProfile retryExecutionProfile = context.getConfig().getProfile(profileName);

        maxRetryCount = retryExecutionProfile.getInt(AmazonKeyspacesRetryOption.KEYSPACES_RETRY_MAX_ATTEMPTS, AmazonKeyspacesRetryOption.DEFAULT_KEYSPACES_RETRY_MAX_ATTEMPTS);

        this.logPrefix = (context != null ? context.getSessionName() : null) + "|" + profileName;
    }


    protected RetryDecision determineRetryDecision(int retryCount) {
        if (retryCount < maxRetryCount) {
            return RetryDecision.RETRY_SAME;
        } else {
            return RetryDecision.RETHROW;
        }
    }

    /**
     * {@inheritDoc}
     *
     * <p>This implementation triggers a maximum of configured retry (to the same connection)
     *
     * <p>Otherwise, the exception is rethrown.
     */
    @Override
    public RetryDecision onReadTimeout(
            @NonNull Request request,
            @NonNull ConsistencyLevel cl,
            int blockFor,
            int received,
            boolean dataPresent,
            int retryCount) {

        RetryDecision decision = determineRetryDecision(retryCount);

        LOG.trace(RETRYING_ON_READ_TIMEOUT, logPrefix, cl, blockFor, received, false, retryCount);

        return decision;

    }


    /**
     * {@inheritDoc}
     *
     * <p>This implementation triggers a maximum of configured retry (to the same connection)
     *
     * <p>Otherwise, the exception is rethrown.
     */
    @Override
    public RetryDecision onWriteTimeout(
            @NonNull Request request,
            @NonNull ConsistencyLevel cl,
            @NonNull WriteType writeType,
            int blockFor,
            int received,
            int retryCount) {

        RetryDecision decision = determineRetryDecision(retryCount);

        LOG.trace(RETRYING_ON_WRITE_TIMEOUT, logPrefix, cl, blockFor, received, false, retryCount);

        return decision;
    }

    /**
     * {@inheritDoc}
     *
     * <p>This implementation triggers a maximum of configured retry (to the same connection)
     *
     * <p>Otherwise, the exception is rethrown.
     */
    @Override
    public RetryDecision onUnavailable(
            @NonNull Request request,
            @NonNull ConsistencyLevel cl,
            int required,
            int alive,
            int retryCount) {

        RetryDecision decision = determineRetryDecision(retryCount);

        LOG.trace(RETRYING_ON_UNAVAILABLE, logPrefix, cl, required, alive, retryCount);

        return decision;
    }

    /**
     * {@inheritDoc}
     *
     * <p>This implementation triggers a maximum of configured retry (to the same connection)
     */
    @Override
    public RetryDecision onRequestAborted(
            @NonNull Request request, @NonNull Throwable error, int retryCount) {

        RetryDecision decision = determineRetryDecision(retryCount);

        LOG.trace(RETRYING_ON_ABORTED, logPrefix, retryCount, error);

        return decision;
    }

    /**
     * {@inheritDoc}
     *
     * <p>This implementation triggers a maximum of configured retry (to the same connection)
     */
    @Override
    public RetryDecision onErrorResponse(
            @NonNull Request request, @NonNull CoordinatorException error, int retryCount) {

        RetryDecision decision = determineRetryDecision(retryCount);

        LOG.trace(RETRYING_ON_ERROR, logPrefix, retryCount, error);

        return decision;
    }

    @Override
    public void close() {
        // nothing to do
    }
}

File: nodejs/datastax-v4/connection-sigv4/AmazonKeyspacesRetryPolicy.js
/**
 * The default implementation triggers a retry on the next host in the query plan with the same consistency level, for historical reasons.
 * 
 * This is a conservative retry policy adapted for the Amazon Keyspaces Service.
 * It allows for a configurable number of attempts, but by default the number of attempts is {@value max_retry_count#default_retry_count}
 * <p>
 * This policy will either reattempt request on the same host or rethrow the exception to the calling thread. The main difference between
 * this policy from the original {@link cassandra.policies.retry.RetryPolicy} is that the {@link AmazonKeyspacesRetryPolicy} will call {@link useCurrentHost} 
 * <p>
 * In Amazon Keyspaces, it's likely that {@link WriteTimeoutException} or {@link ReadTimeoutException} is the result of exceeding current table
 * capacity. Learn more about Amazon Keyspaces capacity here: @see <a href="https://docs.aws.amazon.com/keyspaces/latest/devguide/ReadWriteCapacityMode.html">Amazon Keyspaces CapacityModes</a>.
 * In most cases you should allow for small number of retries, and handle the exception in your application threads.
 *
 * <p>To activate this policy, modify the {@code client connection} section in the driver
 * configuration, for example:
 *
const client = new cassandra.Client({
                   contactPoints: ['cassandra.us-east-2.amazonaws.com'],
                   localDataCenter: 'us-east-2',
                   policies: { retry: new retry.AmazonKeyspacesRetryPolicy(Max_retry_attempts) },
                   pooling: { coreConnectionsPerHost: { [types.distance.local]: 2 } },
                   queryOptions: { isIdempotent: true, consistency: cassandra.types.consistencies.localQuorum },
                   authProvider: auth,
                   sslOptions: sslOptions1,
                   protocolOptions: { port: 9142 }
        });

 * </pre>
 */


'use strict';
const util = require('util');
const cassandra = require('cassandra-driver');
const default_retry_count = 3 // default number of retries on same host
const RetryPolicy = cassandra.policies.retry.RetryPolicy
        
/** @module policies/retry */
/**
* Custom AmazonKeyspacesRetryPolicy.
* Determines what to do when the drivers runs into an specific Cassandra exception
* @constructor
*/
function AmazonKeyspacesRetryPolicy( max_retry_count ) {
    if (max_retry_count) {
                this.retry_count = max_retry_count;
        }
    else{
                this.retry_count = default_retry_count;
        }
}
        
util.inherits (AmazonKeyspacesRetryPolicy, RetryPolicy);
        
/**
* Determines what to do when the driver gets an UnavailableException response
* @param {OperationInfo} info
* @param {Number} consistency The [consistency]{@link module:types~consistencies} level of the query that triggered
* the exception.
* @param {Number} required The number of replicas whose response is required to achieve the
* required [consistency]{@link module:types~consistencies}.
* @param {Number} alive The number of replicas that were known to be alive when the request had been processed
* (since an unavailable exception has been triggered, there will be alive &lt; required)
* @returns {DecisionInfo}
*/
AmazonKeyspacesRetryPolicy.prototype.onUnavailable = function (info, consistency, required, alive) {
    if (info.nbRetry > this.retry_count) {
                return this.rethrowResult();
        }

        return this.retryResult(consistency, true);
};
        
/**
* Determines what to do when the driver gets a ReadTimeoutException response 
* @param {OperationInfo} info
* @param {Number} consistency The [consistency]{@link module:types~consistencies} level of the query that triggered
* the exception.
* @param {Number} received The number of nodes having answered the request.
* @param {Number} blockFor The number of replicas whose response is required to achieve the
* required [consistency]{@link module:types~consistencies}.
* @param {Boolean} isDataPresent When <code>false</code>, it means the replica that was asked for data has not responded.
* @returns {DecisionInfo}
*/
AmazonKeyspacesRetryPolicy.prototype.onReadTimeout = function (info, consistency, received, blockFor, isDataPresent) {
    if (info.nbRetry > this.retry_count) {
        return this.rethrowResult();
    }
        return this.retryResult(consistency, true);
};
        
/**
* Determines what to do when the driver gets a WriteTimeoutException 
* @param {OperationInfo} info
* @param {Number} consistency The [consistency]{@link module:types~consistencies} level of the query that triggered
* the exception.
* @param {Number} received The number of nodes having acknowledged the request.
* @param {Number} blockFor The number of replicas whose acknowledgement is required to achieve the required
* [consistency]{@link module:types~consistencies}.
* @param {String} writeType A <code>string</code> that describes the type of the write that timed out ("SIMPLE"
* / "BATCH" / "BATCH_LOG" / "UNLOGGED_BATCH" / "COUNTER").
* @returns {DecisionInfo}
*/
AmazonKeyspacesRetryPolicy.prototype.onWriteTimeout = function (info, consistency, received, blockFor, writeType) {
    if (info.nbRetry > this.retry_count) {
        return this.rethrowResult();
    }
    return this.retryResult(consistency, true);
};
        
/**
* Defines whether to retry and at which consistency level on an unexpected error.
* <p>
* @param {OperationInfo} info
* @param {Number|undefined} consistency The [consistency]{@link module:types~consistencies} level of the query that triggered
* the exception.
* @param {Error} err The error that caused this request to fail.
* @returns {DecisionInfo}
*/
AmazonKeyspacesRetryPolicy.prototype.onRequestError = function (info, consistency, err) {
    if (info.nbRetry > this.retry_count) {
        return this.rethrowResult();
    }
    return this.retryResult(consistency, true);
};
        
/**
* Returns a {@link DecisionInfo} to retry the request with the given [consistency]{@link module:types~consistencies}.
* @param {Number|undefined} [consistency] When specified, it retries the request with the given consistency.
* @param {Boolean} [useCurrentHost] When specified, determines if the retry should be made using the same host.
* Default: true.
* @returns {DecisionInfo}
*/
AmazonKeyspacesRetryPolicy.prototype.retryResult = function (consistency, useCurrentHost) {
    return {
            decision: RetryPolicy.retryDecision.retry,
            consistency: consistency,
            useCurrentHost: useCurrentHost !== false
        };
};
        
/**
* Returns a {@link DecisionInfo} to callback in error when a err is obtained for a given request.
* @returns {DecisionInfo}
*/
AmazonKeyspacesRetryPolicy.prototype.rethrowResult = function () {
          return { decision: AmazonKeyspacesRetryPolicy.retryDecision.rethrow };
};
        
/**
* Determines the retry decision for the retry policies.
* @type {Object}
* @property {Number} rethrow
* @property {Number} retry
* @property {Number} ignore
* @static
*/
AmazonKeyspacesRetryPolicy.retryDecision = {
    rethrow:  0,
    retry:    1,
    ignore:   2
};
        
exports.AmazonKeyspacesRetryPolicy = AmazonKeyspacesRetryPolicy;